openssl-dynamic/src/main/c/ocsp.c

```diff
@@ -0,0 +1,225 @@
+/*
+ * Copyright 2017 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include "tcn.h"
+#include "ssl_private.h"
+
+// OCSP stapling should be present in OpenSSL as of version 1.0.0
+#if OPENSSL_VERSION_NUMBER < 0x10002000L
+#define TCN_OCSP_NOT_SUPPORTED
+#endif
+
+#if !defined(OPENSSL_NO_OCSP) && !defined(TCN_OCSP_NOT_SUPPORTED) && !defined(OPENSSL_IS_BORINGSSL)
+
+static const int OCSP_CLIENT_ACK = 1;
+static const int OCSP_SERVER_ACK = SSL_TLSEXT_ERR_OK;
+
+/**
+ * This function is being called from OpenSSL. We do everything in
+ * Java-land and this callback is just a stub that returns the
+ * right values to keep OpenSSL happy.
+ *
+ * The arg that is passed into this function is the pointer for one
+ * of these values:
+ *   OCSP_CLIENT_ACK
+ *   OCSP_SERVER_ACK
+ */
+static int openssl_ocsp_callback(SSL *ssl, void *arg) {
+    return *(const int*)arg;
+}
+
+#endif /* !defined(OPENSSL_NO_OCSP) && !defined(TCN_OCSP_NOT_SUPPORTED) && !defined(OPENSSL_IS_BORINGSSL) */
+
+/**
+ * Enables OCSP stapling for the given SSLContext
+ */
+TCN_IMPLEMENT_CALL(void, SSLContext, enableOcsp)(TCN_STDARGS, jlong ctx, jboolean client) {
+
+    tcn_ssl_ctxt_t *c = J2P(ctx, tcn_ssl_ctxt_t *);
+    if (c == NULL) {
+        tcn_ThrowException(e, "ctx is null");
+        return;
+    }
+
+#if defined(OPENSSL_NO_OCSP)
```

Comment:
so the simplistic approach I had in mind is:

`#if defined(OPENSSL_NO_OCSP)` -> `#if defined(OPENSSL_NO_OCSP) && !defined(OPENSSL_IS_BORINGSSL)`

We may be able to refactor the macros a bit but basically if `OPENSSL_IS_BORINGSSL == true` we should just ignore `OPENSSL_NO_OCSP` and use the BoringSSL APIs.


```diff
diff --git a/openssl-dynamic/src/main/c/ocsp.c b/openssl-dynamic/src/main/c/ocsp.c
index 317490e..f625817 100644
--- a/openssl-dynamic/src/main/c/ocsp.c
+++ b/openssl-dynamic/src/main/c/ocsp.c
@@ -54,7 +54,7 @@ TCN_IMPLEMENT_CALL(void, SSLContext, enableOcsp)(TCN_STDARGS, jlong ctx, jboolea
         return;
     }
 
-#if defined(OPENSSL_NO_OCSP)
+#if defined(OPENSSL_NO_OCSP) && !defined(OPENSSL_IS_BORINGSSL)
     tcn_ThrowException(e, "netty-tcnative was built without OCSP support");
 
 #elif defined(TCN_OCSP_NOT_SUPPORTED)
@@ -107,7 +107,7 @@ TCN_IMPLEMENT_CALL(void, SSL, enableOcsp)(TCN_STDARGS, jlong ssl) {
         return;
     }
 
-#if defined(OPENSSL_NO_OCSP)
+#if defined(OPENSSL_NO_OCSP) && !defined(OPENSSL_IS_BORINGSSL)
     tcn_ThrowException(e, "netty-tcnative was built without OCSP support");
 
 #elif defined(TCN_OCSP_NOT_SUPPORTED)
@@ -139,14 +139,14 @@ TCN_IMPLEMENT_CALL(void, SSL, setOcspResponse)(TCN_STDARGS, jlong ssl, jbyteArra
         return;
     }
 
-#if defined(OPENSSL_NO_OCSP)
+#if defined(OPENSSL_NO_OCSP) && !defined(OPENSSL_IS_BORINGSSL)
     tcn_ThrowException(e, "netty-tcnative was built without OCSP support");
 
 #elif defined(TCN_OCSP_NOT_SUPPORTED)
     tcn_ThrowException(e, "OCSP stapling is not supported");
 
 #elif defined(OPENSSL_IS_BORINGSSL)
-    const uint8_t *value = OPENSSL_malloc(sizeof(uint8_t) * length);
+    uint8_t *value = OPENSSL_malloc(sizeof(uint8_t) * length);
     if (value == NULL) {
         tcn_ThrowException(e, "OPENSSL_malloc() returned null");
         return;
@@ -192,7 +192,7 @@ TCN_IMPLEMENT_CALL(jbyteArray, SSL, getOcspResponse)(TCN_STDARGS, jlong ssl) {
         return NULL;
     }
 
-#if defined(OPENSSL_NO_OCSP)
+#if defined(OPENSSL_NO_OCSP) && !defined(OPENSSL_IS_BORINGSSL)
     tcn_ThrowException(e, "netty-tcnative was built without OCSP support");
 
 #elif defined(TCN_OCSP_NOT_SUPPORTED)

```